lib/leanplum_api/api.rb

```diff
@@ -1,58 +1,53 @@
 module LeanplumApi
   class API
+    EXPORT_PENDING = 'PENDING'.freeze
+    EXPORT_RUNNING = 'RUNNING'.freeze
+    EXPORT_FINISHED = 'FINISHED'.freeze
     extend Gem::Deprecate
 
     class LeanplumValidationException < RuntimeError; end
 
-    EXPORT_PENDING = 'PENDING'
-    EXPORT_RUNNING = 'RUNNING'
-    EXPORT_FINISHED = 'FINISHED'
-
     def initialize(options = {})
       fail 'LeanplumApi not configured yet!' unless LeanplumApi.configuration
     end
 
     def set_user_attributes(user_attributes, options = {})
-      track_multi(nil, user_attributes, options)
+      track_multi(user_attributes: user_attributes, options: options)
+    end
+
+    def set_device_attributes(device_attributes, options = {})
+      track_multi(device_attributes: device_attributes, options: options)
     end
 
     def track_events(events, options = {})
-      track_multi(events, nil, options)
+      track_multi(events: events, options: options)
     end
 
-    # This method is for tracking events and/or updating user attributes at the same time, batched together like
-    # leanplum recommends.
-    # Set the :force_anomalous_override option to catch warnings from leanplum about anomalous events and force them to
-    # not be considered anomalous.
-    def track_multi(events = nil, user_attributes = nil, options = {})
-      request_data = Array.wrap(user_attributes).map { |h| build_user_attributes_hash(h) } +
-                     Array.wrap(events).map { |h| build_event_attributes_hash(h, options) }
+    # This method is for tracking events and/or updating user attributes
+    # at the same time, batched together like leanplum recommends.
+    # Set the :force_anomalous_override option to catch warnings from leanplum
+    # Ôabout anomalous events and force them to not be considered anomalous.
+    def track_multi(events: nil, user_attributes: nil, device_attributes: nil, options: {})
+      events = Array.wrap(events)
+
+      request_data = []
+      if user_attributes.present?
```

Comment:
you don't need this conditional.  a `nil` will just not add anything to the `request_data`
```ruby
irb(main):002:0> Array.wrap(nil)
=> []
```